File: C++_version/math_tools.cpp
#include "math_tools.h"
// #include <vector>

typedef long double ld;

// Time: O(n)
// Space: O(1)
ld math_tools::get_quad_nth_v0(vector<ld>S, int posn) {
    // should be globals for α, β and x to avoid resourec duplication, or make them part of the class but only accessible by this function
    ld α = S[1] - S[0];
    ld β = S[2] - S[1];
    ld x = β - α;
    ld sum = 0.0;
    // such loops will likely increase depending on the magnitude of the series. Could be solved with recursion (in polynomial time)
    for(int i = 1; i < posn; i++)
        sum += ld((x*i)+(α-x));
    return ld(S[0]+sum);
}

// Time: O(n)
// Space: O(1)
ld math_tools::parr_res_v0(vector<ld>Ω) {
    ld reciprocal_sum = 0.0;
    for(int i = 0; i < Ω.size(); i++)
        reciprocal_sum += ld(1/Ω[i]);
    return ld(1/reciprocal_sum);
}

File: Python_version/math_tools.py
class my_math:

    # Time: O(n)
    # Space: O(1)
    def get_quad_nth_v0(self, S: list, n: int) -> float:
        # need these as attributes, since I need to access them for printing the formula
        # make them part of the class but only accessible by this function
        α = S[1] - S[0]
        β = S[2] - S[1]
        x = β - α
        sum = 0.0
        # such loops will likely increase depending on the magnitude of the series. Could be solved with recursion (in polynomial time)
        for i in range(1, n):
            sum += float((x*i)+(α-x))
        return float(S[0]+sum)

    # Time: O(n)
    # Space: O(1)
    def parr_res_v0(self, Ω: list) -> float:
        recoprocal_sum = 0.0
        for i in range(len(Ω)):
            recoprocal_sum += float(1/Ω[i])
        return float(1/recoprocal_sum)